pipeline_blocks.py

```diff
@@ -370,12 +419,202 @@ def _groupby_aggregations(dispatchers, config, train_mode, **kwargs):
+                                                       transformer=previous_applications_groupby_agg,
+                                                       input_data=['input'],
+                                                       adapter=Adapter({'X': E('input', 'X_valid'),
+                                                                        'file': E('input', 'previous_application')}),
+                                                       experiment_directory=config.pipeline.experiment_directory,
+                                                       **kwargs)
+
+        return previous_applications_groupby_agg, previous_applications_groupby_agg_valid
+
+    else:
+        previous_applications_groupby_agg = Step(name='previous_applications_groupby_agg',
+                                                 transformer=fe.GroupbyAggregationFromFile(**config.previous_applications),
+                                                 input_data=['input'],
+                                                 adapter=Adapter({'X': E('input', 'X'),
+                                                                  'file': E('input', 'previous_application')}),
+                                                 experiment_directory=config.pipeline.experiment_directory,
+                                                 **kwargs)
+
+        return previous_applications_groupby_agg
+
+
+def _application(config, train_mode, **kwargs):
+    if train_mode:
+        application = Step(name='application',
+                           transformer=fe.ApplicationFeatures(),
+                           input_data=['input'],
+                           adapter=Adapter({'X': E('input', 'X')}),
+                           experiment_directory=config.pipeline.experiment_directory,
+                           **kwargs)
+
+        application_valid = Step(name='application_valid',
+                                 transformer=application,
+                                 input_data=['input'],
+                                 adapter=Adapter({'X': E('input', 'X_valid')}),
+                                 experiment_directory=config.pipeline.experiment_directory,
+                                 **kwargs)
+
+        return application, application_valid
+
+    else:
+        application = Step(name='application',
+                           transformer=fe.ApplicationFeatures(),
+                           input_data=['input'],
+                           adapter=Adapter({'X': E('input', 'X')}),
+                           experiment_directory=config.pipeline.experiment_directory,
+                           **kwargs)
+
+        return application
+
+
 def _bureau(config, train_mode, **kwargs):
     if train_mode:
         bureau = Step(name='bureau',
-                      transformer=fe.GroupbyAggregationFromFile(**config.bureau),
+                      transformer=fe.BureauFeatures(**config.bureau),
                       input_data=['input'],
```

Comment:
@pknut I think it would be better to have input_data=['bureau','main'] and then in adapter X: 'main', 'X and bureau: 'bureau', 'X' . It shows the benefit of multiinput  data dictionary that is passed to step in the main.py file